### ISSUE TITLE
Delays ?

### ISSUE THREAD

**Issue by rafaqz** [cesaraustralia/DynamicGrids.jl#110]
Delays could be very useful to have in rules, and seeing we already often store the history of the simulation, not very difficult to achieve.

@jamesmaino I was thinking about  reproduction rates in your long live insect and how the growth rate should lag the current population (unless you have a size-structured population). 

We could use previous populations as a parameter, like `Grid` and `Aux` but like `Delay{:gridname, 3)` for 3 timesteps before the current one - or explicitly use a multiple of the simulation timestep `Delay{:gridname, Month(3))`.

Anyway, an idea for later, just recording it here.

**Comment by jamesmaino** [comment]
This would be useful. But not sure I understand your method. 

I get that we can used the stored arrays for previous timesteps, which is a nice idea (i.e. no change if delayed), but how to we know when a delay for a gridcell is active in the current timestep (i.e. what triggers the rule)? And then how do we skip some (or all) other operations for efficiency (e.g. skip `local_dispersal`, `growth`, etc).

**Comment by rafaqz** [comment]
I'm not sure we have to worry about efficiency, this isn't really any different to using `Aux{:auxname}()` or `Grid{:gridname}()` to get parameters. It will just get the parameter for the current cell N steps back, in the specified grid, if you use `Delay{:gridname,3}()` for the parameter value.

We always know the current timestep/frame number with `currenttime(simdata)` and `currentframe(simdata)` so we would just count back N steps. `Delay{:gridname, 3}` would get use the value of the `:gridname` grid 3 steps back.

It would mean a method like this:

```julia
get(simdata::SimData, ::Delay{X,N}, I...) where {X, N} = ... 
```

And for `...` we retrieve the value of grid `X` at `currentframe(simdata) - N` and indices `I`. The only thing missing for this to only be a few lines of code is adding a link to the `Output` frames in the `SimData` object, and checking to make sure `isstored == true`, which would be a conditon of using a `Delay`.

**Comment by jamesmaino** [comment]
Oh right. Yeah - that could work. Not just efficiency though; what if you want to stop a cell dispersing? E.g. other cells can keep growing and dispersing, but current cell is frozen. This would only work for `OutwardsDispersal`.

**Comment by rafaqz** [comment]
I'm not sure I get the connection between that and delays. Do you mean they shouldn't disperse based on a delayed value? Dispersal with a time-delayed threshold? 

You would have to code that rule. But it would be nice if we had a `Condition` wrapper like `Chain` that wrapped an existing rule and runs it only if some condition is met.

**Comment by jamesmaino** [comment]
Here is a use case: 

An insect goes into diapause in certain grid cells where it is cold. For grid cells where insects are in diapause, populations will not grow, disperse, but may still suffer some mortality, and may see incoming migrants from other cells not in diapause.

**Comment by rafaqz** [comment]
Ok, makes sense. Not sure you need the delays for this though, I think its a different problem. This is where we need that `Condition` wrapper to avoid having to rewrite all the other rules to check for diapause.

I would define another grid of `Bool` called  `:diapause`, and have:

```julia
conditional_rules = Condition{:diapause}(Chain(disperse, grow)) do data, diapause, I
     diapause
end
``` 
And it would only run the contained rule if it met the condition - and here diapause is already a `Bool`

(Or instead have 2 separate `Condition` wappers for dispersal and growth, as actually chaining neighborhood rules with the aux data growth rule makes them slower now! its too fast and optimized now, so it's easy to slow back down to previous speeds - minor additions to code page size and pulling in extra data break how efficient the kernel convolution is. but anyway...)

**Comment by jamesmaino** [comment]
Great. I was going to make another issue for conditional rules but you beat me to it! :smiley:

**Comment by rafaqz** [comment]
Yeah I think the general idea of wrapper rules could add another layer of flexibility to everything. Just need to abstract out the code from `Chain` a little more.

Let me know if you have any better syntax ideas for `Condition` (or a better name). We could even have `If` and `IfElse` to do switching between rules.

**Comment by rafaqz** [comment]
Closed by #157
